scality_manila_utils/smb_helper.py

```diff
@@ -0,0 +1,351 @@
+    def wrapper(root_export, *args, **kwargs):
+        verify_environment(root_export)
+        return f(root_export=root_export, *args, **kwargs)
+
+    return wrapper
+
+
+def ensure_export_exists(f):
+    """
+    Decorator function which verifies that a given export exists and pass
+    the `dict` of all defined exports to the decorated function.
+    """
+    @functools.wraps(f)
+    def wrapper(export_name, *args, **kwargs):
+        exports = _get_defined_exports()
+        if export_name not in exports:
+            msg = "Share '{0:s}' not found in Samba registry.".format(
+                  export_name)
+            raise ExportNotFoundException(msg)
+
+        return f(export_name=export_name, exports=exports, *args, **kwargs)
+
+    return wrapper
+
+
+@ensure_environment
+@ensure_export_exists
+def get_export(export_name, exports, *args, **kwargs):
+    """
+    Retrieve client details of an export.
+
+    :param export_name: name of export
+    :type export_name: string (unicode)
+    :param exports: all the defined shares in the Samba registry
+    :type exports: dictionary
+    :returns: string with export client details in json format
+    """
+
+    export = exports[export_name]
+    clients = dict((host, ["rw"]) for host in export['hosts allow'].split())
+
+    return json.dumps(clients)
+
+
+@ensure_environment
+def add_export(root_export, export_name, *args, **kwargs):
+    """
+    Add an export.
+
+    :param root_export: SOFS directory which holds the export points exposed
+        through manila
+    :type root_export: string (unicode)
+    :param export_name: name of export to add
+    :type export_name: string (unicode)
+    """
+
+    if not export_name or '/' in export_name:
+        raise ExportException('Invalid export name')
+
+    export_point = os.path.join(root_export, export_name)
+
+    create_cmd = [
+        'net', 'conf', 'addshare', export_name, export_point,
+        'writeable=y', 'guest_ok=y',
+    ]
+    parameters = {
+        'browseable': 'yes',
+        'create mask': '0755',
+        'hosts deny': '0.0.0.0/0',  # deny all by default
+        'hosts allow': '127.0.0.1',
+        'read only': 'no',
+    }
+
+    set_of_commands = [['net', 'conf', 'setparm', export_name,
+                        param, value] for param, value in parameters.items()]
+
+    with utils.elevated_privileges():
+        try:
+            os.mkdir(export_point)
+        except OSError as exc:
+            if exc.errno != errno.EEXIST:
+                raise
+            else:
+                log.debug("The share/directory %s already exists on SOFS",
+                          export_name)
+                exports = _get_defined_exports()
+                if export_name in exports:
+                    msg = ("Share '{0:s}' already defined in Samba "
+                           "registry.".format(export_name))
+                    raise ExportAlreadyExists(msg)
+
+        os.chmod(export_point, 0o0777)
```

Comment:
I want the `add_export` function to be idempotent, somehow... I probably failed at this.